*Problem Statement* :
<br>4.	RPL Banking and Financing company wants to study the behavior patterns of their customers so that they can efficiently provide their services and solve the problem of churn. They have historical data of their customers. Build an Artificial Neural Network with Exited as the target variable.

In [1]:
#importing required libraries
import pandas as pd
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense 
from tensorflow.keras.callbacks import EarlyStopping
import numpy as np

In [2]:
#declaring seeds to reproduce same resullts
seeds=7
np.random.seed(seeds)

In [4]:
#designing model
def design_model():
    model= Sequential()
    model.add(Dense(12,input_dim=15 , kernel_initializer="uniform",activation="relu"))
    model.add(Dense(24,kernel_initializer="uniform",activation="relu"))
    model.add(Dense(24,kernel_initializer="uniform",activation="relu"))
    model.add(Dense(6,kernel_initializer="uniform",activation="relu"))
    model.add(Dense(1,kernel_initializer="uniform",activation="sigmoid"))
    model.compile(loss="binary_crossentropy",optimizer="adam",metrics=["accuracy"])
    return model

In [5]:
model=design_model()
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 12)                192       
_________________________________________________________________
dense_1 (Dense)              (None, 24)                312       
_________________________________________________________________
dense_2 (Dense)              (None, 24)                600       
_________________________________________________________________
dense_3 (Dense)              (None, 6)                 150       
_________________________________________________________________
dense_4 (Dense)              (None, 1)                 7         
Total params: 1,261
Trainable params: 1,261
Non-trainable params: 0
_________________________________________________________________


In [6]:
#loading data set
dt=pd.read_csv("file:///D:/DataScience/Class/assignment working/ANN/RPL.csv")    
dt.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [7]:
#cheking null values
dt.isna().sum()

RowNumber          0
CustomerId         0
Surname            0
CreditScore        0
Geography          0
Gender             0
Age                0
Tenure             0
Balance            0
NumOfProducts      0
HasCrCard          0
IsActiveMember     0
EstimatedSalary    0
Exited             0
dtype: int64

In [8]:
#creating dummies for categorical data and droping unwanted veriable
dt=dt.drop("Surname",axis=1)
data=pd.get_dummies(dt,columns=["Geography","Gender"])

In [9]:
#Separating Target and input variables
X=data.drop("Exited",axis=1)
X=np.array(X)
Y=data.Exited
Y=np.array(Y)

convering into array to make computation faster

In [10]:
#applying monitor to stop training if there is no considerable change in validation loss 
monitor=EarlyStopping("val_loss",min_delta=1e-3,patience=5,verbose=1,mode="auto",restore_best_weights=True)


In [11]:
#fitting model
model.fit(X,Y, validation_split=0.25, epochs=100 ,batch_size=10,use_multiprocessing=True,callbacks=[monitor])

Epoch 1/100
750/750 [==============================] - 2s 3ms/step - loss: 0.5581 - accuracy: 0.7927 - val_loss: 0.4948 - val_accuracy: 0.8036
Epoch 2/100
750/750 [==============================] - 2s 2ms/step - loss: 0.5209 - accuracy: 0.7939 - val_loss: 0.5029 - val_accuracy: 0.8036
Epoch 3/100
750/750 [==============================] - 2s 3ms/step - loss: 0.5152 - accuracy: 0.7939 - val_loss: 0.5146 - val_accuracy: 0.8036
Epoch 4/100
750/750 [==============================] - 2s 2ms/step - loss: 0.5154 - accuracy: 0.7939 - val_loss: 0.5044 - val_accuracy: 0.8036
Epoch 5/100
750/750 [==============================] - 2s 2ms/step - loss: 0.5120 - accuracy: 0.7939 - val_loss: 0.4940 - val_accuracy: 0.8036
Epoch 6/100
750/750 [==============================] - 1s 2ms/step - loss: 0.5122 - accuracy: 0.7939 - val_loss: 0.4901 - val_accuracy: 0.8036
Epoch 7/100
750/750 [==============================] - 2s 3ms/step - loss: 0.5106 - accuracy: 0.7939 - val_loss: 0.4898 - val_accuracy: 0.8036

We can see model stopped after 14th epoch as there was no improvement in validation accuracy

* Model is working fine and giving acceptablle accuracy
* No overfitting
* 80% validation accuracy and 79.39% testing accuracy